### ISSUE TITLE
Make concat no-op. 

### ISSUE THREAD

**Issue by wujingyue** [NVIDIA/Fuser#1502]
While it's in general beneficial for nvFuser to generate a fast concat kernel, sometimes concats can be made a no-op when we allocate tensors carefully. Thanks @tfogal for mentioning this opportunity. 

Here's a pattern from the nanogpt gpt2 grad benchmark: 

```
t6698, t6699, t6700 = scaled_dot_product_attention_backprop(...)  # bf16[16,12,128,64], bf16[16,12,128,64], bf16[16,12,128,64]

t6703 = transpose(t6698, [0, 2, 1, 3])  # [16, 128, 12, 64]
t6704 = reshape(t6703, [16, 128, 768])  # [16, 128, 768]
t6705 = transpose(t6699, [0, 2, 1, 3])  # [16, 128, 12, 64]
t6706 = reshape(t6705, [16, 128, 768])  # [16, 128, 768]
t6701 = transpose(t6700, [0, 2, 1, 3])  # [16, 128, 12, 64]
t6702 = reshape(t6701, [16, 128, 768])  # [16, 128, 768]

t6707 = concat(t6704, t6706, t6702, axis=-1)  # [16, 128, 2304]

t6715 = cast(t6707, f32)  # [16, 128, 2304]
t6716 = sum(t6715, [0, 1])  # [2304]
# Output 0
t6717 = cast(t6716, bf16)  # [2304]
# Output 1
t6708 = view(t6707, [2048, 2304])
# Output 2
t6712 = transpose(t6708, [1, 0])  # [2034, 2048]
```

To me, the hardest piece of the puzzle is buffer allocation. The best outcome is to make the concat's input buffers slices of its output buffer. The current buffer allocation algorithm in `csrc/executor.cpp` follows the execution order, so it has to allocate the concat's input buffers (i.e. producers' output buffers) before the output. 

Orthogonally, some code motion may speed things up but doesn't solve the above buffer allocation problem. @jjsjann123 mentioned the "unzipping" idea, which moves the concat close to the fusion outputs. We could also do "zipping", which moves up the concat to right after the three fusion inputs. This works even better if `scaled_dot_product_attention_backprop` can output to the desired layout with good performance. Apparently, the `mha_bwd` kernel only requires [the feature dimension to be contiguous](https://github.com/pytorch/pytorch/blob/670eb83573cb69558b2149079eb2a3c74bdd5c4e/aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.cpp#L782). 

cc @naoyam @jjsjann123 @kevinstephano @protonu @tfogal @Priya2698

**Comment by wujingyue** [comment]
> This works even better if scaled_dot_product_attention_backprop can output to the desired layout with good performance. Apparently, the mha_bwd kernel only requires [the feature dimension to be contiguous](https://github.com/pytorch/pytorch/blob/670eb83573cb69558b2149079eb2a3c74bdd5c4e/aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.cpp#L782).

cc @rdspring1 who might know about this!

**Comment by rdspring1** [comment]
Yes, the fused sdpa kernels want the query, key, and value tensors to be permuted `[0, 2, 1, 3]`. The only requirement is for the inner-most dimension to be contiguous.

**Comment by wujingyue** [comment]
> the fused sdpa kernels want the query, key, and value tensors to be permuted [0, 2, 1, 3]

Are you saying the fused kernel **requires** the output tensor to be allocated in this certain stride order?

**Comment by jjsjann123** [comment]
sounds like Ryan is talking about input rather than outputs....

**Comment by wujingyue** [comment]
Inside `_scaled_dot_product_efficient_attention_backward`, `dq`, `dk` and `dv` are allocated separately according to https://github.com/pytorch/pytorch/blob/6243dbb5c08e9b3f87bc7dc74816b2ceb6957a94/aten/src/ATen/native/transformers/cuda/attention_backward.cu#L236-L238. We could change the code to allocate one buffer and split it into three. However, this implementation is low-level and needs to handle cases where the query and the key/value have different sequence lengths, so it's not always possible to allocate and split. I won't shut this door completely, but that complexity drives to consider other alternatives...

**Comment by wujingyue** [comment]
Re: the unzipping approach

1. [compiler/optimization] Create a pre-segmenter pass that moves concat to the end and adds segment_set before concat.
2. [compiler/scheduling] The NoOp scheduler accepts a concat-only Fusion and marks its inputs as aliases of the output. 
3. [compiler/lowering] FusionExecutor::compileFusion optionally creates a separate fusion IR for alias computation. This will be embedded in the FusionExecutor as a separate IrContainer. This is used to represent the aliases from outputs to inputs due to concatenation. As a side benefit, this can also be used to represent aliases of type InplaceUpdate – the allocation IR describes allocation and the existing fusion describes computation. Generating such an IR needs to know the size of each slice, which is only available through the original inputs. One possible solution is to let FusionExecutor take meta tensors indicating the slice sizes. 
4. [runtime/allocation] Hoist input/output buffer allocation from runFusion to FusionKernelRuntime. Temporary buffers will still be allocated by runFusion. This way, we can reuse a preallocated buffer as the output buffer, instead of (as is) always allocating the output buffer. 
5. [runtime/execution] 
   ```
   For each segment group in the run order:
     Retrieve the executor and the underlying FusionExecutor::kernel().
     For each input/output of the segment group:
       If it’s already mapped to a tensor, use that tensor.
       If it’s not an (input or output) alias, allocate it.
       If it is an alias, allocate the root and compute the alias. In the future, we may need this to happen recursively, but just one look-ahead seems to be enough for this particular case of concat. 
     runFusion with allocated/computed inputs and outputs.
   ```

I'm sure this is one of the many ways, so let me know your thoughts! This may be too big a hammer to tackle this one particular concat use case. Some smaller hammers I considered either accumulate huge tech debts or aren't much simpler to implement.

**Comment by wujingyue** [comment]
https://github.com/NVIDIA/Fuser/commit/8ea3f1e148ae8fdc82744bcc5e6f773601679166 shows an idea that avoids removes step 3 and makes step 2 slightly more complicated: instead of creating a separate IR, we can let the NoOp scheduler append an alias-computing epilogue to the fusion.

**Comment by wujingyue** [comment]
I realized I have made this thread a mini design doc 😄. Let me try to summarize my thoughts so far. 

The two use case: 
1. [the pattern found in nanoGPT](https://github.com/NVIDIA/Fuser/issues/1502#issue-2038300603)
2. the pattern found in LLaMA. See #1597 

The three options to consider:
1. [zipping](https://github.com/NVIDIA/Fuser/issues/1502#issuecomment-1870837878)
2. [segmentation](https://github.com/NVIDIA/Fuser/issues/1502#issuecomment-1871744620)
3. codegen: optimize nvFuser to generate better CUDA kernel for fusions that contain Cat. I'll track this option separately in #1597 

Decision matrix:

| Options      | nanoGPT | LLaMA  | work and risks 
| ------------ | ------- | ------ | --------------
| zipping      | best    | worst  | cudnn's `sdpa_backward` sometimes generates NaNs, which @vedaanta-nvidia is trying to fix. Other work can be done within one month.
| segmentation | worst   | medium | About two months of plumbing in nvFuser.
| codegen      | medium  | best   | @naoyam pointed me to some issues with vectorization (easier) and parallelization (harder). I don't have a clear estimate yet.

Footnote: best/medium/worst is in terms of performance of generated code. 

My recommendation so far: 
1. Don't go with the **segmentation** option. It requires non-trivial work and doesn't lead to optimal code. In the nanoGPT case, it creates a kernel boundary between `Cat` and `Sum` that causes extra data reads. In the LLaMA case, even with some non-trivial code motion that downstreams `Cat`, it still leads to multiple smaller kernels that could be horizontally fused. 
2. Go with the **codegen** option because it leads to the best code for LLaMA and is generally useful in the long term. 
3. Also go with the **zipping** option, the only option among the three that makes concat free (thus, best code) for the nanoGPT use case. Caveat: we'll have to wait for the NaN fix before landing it. Some general improvements can go in in parallel, e.g., upstreaming `Cat` through unary ops.

**Comment by wujingyue** [comment]
Mark this as complete because the zip approach [has been implemented](https://github.com/NVIDIA/Fuser/issues/1768) and that @jjsjann123 and @naoyam are working on the codegen approach (tracked or will be tracked in a separate issue).
